radialdf/__init__.py

```diff
@@ -1,87 +1,230 @@
-    # This does mean that the radial distribution of particles in this border region is
-    # not examined. See #1 for future implementation of the overlap library to fix this.
-    search_box[:, 0] += r
-    search_box[:, 1] -= r
-    # Find all the reference particles that fall within the shrunken bounds and won't
-    ref = _within(particles, search_box)
-    tree = KDTree(particles)
-    # Query the tree for the neighbours within radius r of all the reference particles.
-    all_neighbours = tree.query_ball_point(particles[ref], r=r)
-    # Create an iterator that turns the reference id into the particle id and the
-    # neighbour list into a neighbour np.ndarray
-    nbr_iter = map(lambda x: (ref[x[0]], np.array(x[1])), enumerate(all_neighbours))
-    # A local function that turns neighbour ids into distances and excludes the query
-    # particle with id `me`, using the `particles` of this function call scope.
-    def dist_excl_self(me, my_nbr):
-        offsets = particles[my_nbr[my_nbr != me]] - particles[me]
-        return np.sqrt(np.sum(offsets ** 2, axis=1))
-    # Concatenate all the distances of all the neighbours of all the reference points
-    # together
-    all_distances = np.concatenate(tuple(dist_excl_self(me, my_nbr) for me, my_nbr in nbr_iter))
-    # Bin them in `dr` sized bins
-    radial_bins = np.bincount(np.floor(all_distances / dr).astype(int))
-    # Normalize the result to the amount of reference points were tested
-    trials = len(ref)
-    # Normalize the bins to the volume of the shell that corresponds to the radial bin
-    radii = np.arange(0, len(radial_bins) * dr, dr)
-    inner_volumes = 4 / 3 * np.pi * radii ** 3
-    outer_volumes = 4 / 3 * np.pi * (radii + dr) ** 3
-    bin_volumes = outer_volumes - inner_volumes
-    # Normalize the result to the density of the particles
-    density = len(particles) / np.product(np.abs(np.diff(boundary, axis=1)))
-    # Return the normalized radial distribution
-    return radial_bins / trials / bin_volumes / density
-
-def _within(particles, boundary):
-    within = None
-    for _dim, (low, high) in enumerate(boundary):
-        dim = particles[:, _dim]
-        mask = (dim >= low) & (dim <= high)
-        if within is not None:
-            mask = mask & within
-        within = mask
-    return np.nonzero(within)[0]
-
-def rdf(boundary, particles, r, dr):
-    raise NotImplementedError("The RDF dealing with border effects is not implemented yet, use `inner_rdf` instead, which excludes the border regions.")
+import numpy as np 
+from scipy.spatial import cKDTree
+import functools
+import plotly.graph_objects as go 
+import math
+import itertools
+
+PLOTTING = False
+_pi = math.pi
+
+try:
+    functools.cache
+except AttributeError:
+    functools.cache = functools.lru_cache(None)
+
+def _pairs(iterable):
+    "s -> (0, s0), (s0,s1), (s1,s2), (s2, s3), ..."
+    a, b = itertools.tee(iterable)
+    a0 = next(b, None)
+    yield 0, a0
```

Comment:
This should be reduced to `yield 0, next(b)`:

1. Shorter 😇 
2. If `iterable` is empty a `StopIteration` should be raised, in the current form we'll yield `(0, None)` before a `StopIteration`.